BeeKit/Managers/GoalManager.swift

```diff
@@ -57,17 +57,60 @@ public actor GoalManager {
 
     /// Fetch and return the latest set of goals from the server
     public func refreshGoals() async throws {
-        guard let username = await currentUserManager.username else {
-            try await currentUserManager.signOut()
-            return
+        guard let user = self.currentUserManager.user(context: modelContext) else { return }
+        let goalsUnknown = user.goals.count == 0 || user.updatedAt.timeIntervalSince1970 < 24*60*60
+
+        let userResponse: JSON
+        let goalResponse: JSON
+
+        let deleteMissingGoals: Bool
+        let deletedGoals: JSON
+
+        if (goalsUnknown) {
+            logger.notice("Goals unknown, doing full fetch")
+            // We must fetch the user object first, and then fetch goals afterwards, to guarantee User.updated_at is
+            // a safe timestamp for future fetches without losing data
+            userResponse = JSON(try await requestManager.get(url: "api/v1/users/{username}.json")!)
+            goalResponse = JSON(try await requestManager.get(url: "api/v1/users/{username}/goals.json")!)
+
+            deleteMissingGoals = true
+            deletedGoals = JSON(arrayLiteral: [])
+        } else {
+            // TODO: Use old data to find deleted goals
+            logger.notice("Doing incremental update since \(user.updatedAt, privacy: .public)")
+            userResponse = JSON(try await requestManager.get(url: "api/v1/users/{username}.json", parameters: ["diff_since": user.updatedAt.timeIntervalSince1970 + 1])!)
+            goalResponse = userResponse["goals"]
+
+            deleteMissingGoals = false
+            deletedGoals = userResponse["deleted_goals"]
         }
 
-        let responseObject = try await requestManager.get(url: "api/v1/users/\(username)/goals.json")!
-        let response = JSON(responseObject)
-        self.updateGoalsFromJson(response) // TODO: Return failure info
+        // The user may have logged out during the network operation. If so we have nothing to do
+        modelContext.refreshAllObjects()
+        guard let user = modelContext.object(with: user.objectID) as? User else { return }
 
-        self.goalsFetchedAt = Date()
+        user.updateToMatch(json: userResponse)
```

Comment:
GoalManager now knows how to fetch user and set user in db. What is CurrentUserManager's role?

Reply:
I'm thinking GoalManager should turn into a general SyncManager which is responsible for all writes to CoreData. This will also avoid needing all the data refreshing logic. CurrentUserManager will just handle login/logout.